pkg/llbutil/id.go

```diff
@@ -34,13 +37,31 @@ func LocalID(ctx context.Context, absPath string, opts ...llb.LocalOption) (stri
 	return digest.FromBytes(def.Def[len(def.Def)-1]).String(), nil
 }
 
-// LocalUniqueID returns a consistent string that is unique per host + cwd
-func LocalUniqueID(cwd string) (string, error) {
+// localUniqueID returns a consistent string that is unique per host + dir +
+// last modified time.
+//
+// If there is already a solve in progress using the same local dir, we want to
+// deduplicate the "local" if the directory hasn't changed, but if there has
+// been a change, we must not identify the "local" as a duplicate. Thus, we
+// incorporate the last modified timestamp into the result.
+func localUniqueID(dir string) (string, error) {
 	mac, err := FirstUpInterface()
 	if err != nil {
 		return "", err
 	}
-	return fmt.Sprintf("cwd:%s,mac:%s", cwd, mac), nil
+
+	var lastModified time.Time
+	err = filepath.Walk(dir, func(path string, info fs.FileInfo, err error) error {
+		if lastModified.IsZero() || info.ModTime().After(lastModified) {
```

Comment:
This ignores the llb.LocalOption, so will potentially cause more churn than necessary if excluded files are frequently updated.  I wonder if we can use the include/exclude patterns to get more accurate here?

Reply:
Hmm, I can look into using `fsutil.Walk` instead of `filepath.Walk` here (it honors include/exclude patterns).

Reply:
Updated to use `fsutil.Walk` with include/exclude patterns honored. In my testing this seems to solve the original problem.

Reply:
nice! it looks like the input `dir` might also be a file, so I think we can just short-circuit that scenario.